lib/collection/src/collection_manager/segments_searcher.rs

```diff
@@ -439,37 +422,69 @@ fn search_in_segment(
 
     // run last batch if any
     if !vectors_batch.is_empty() {
-        let locked_segment = segment.get();
-        let read_segment = locked_segment.read();
-        let segment_points = read_segment.available_point_count();
-        let top = if use_sampling {
-            let ef_limit = prev_params
-                .params
-                .and_then(|p| p.hnsw_ef)
-                .or_else(|| get_hnsw_ef_construct(read_segment.config(), prev_params.vector_name));
-            sampling_limit(prev_params.top, ef_limit, segment_points, total_points)
-        } else {
-            prev_params.top
-        };
-        let mut res = read_segment.search_batch(
-            prev_params.vector_name,
+        let (mut res, mut further) = execute_batch_search(
+            &segment,
             &vectors_batch,
-            &prev_params.with_payload,
-            &prev_params.with_vector,
-            prev_params.filter,
-            top,
-            prev_params.params,
+            &prev_params,
+            use_sampling,
+            total_points,
             is_stopped,
         )?;
-        for batch_result in &res {
-            further_results.push(batch_result.len() == top);
-        }
+        further_results.append(&mut further);
         result.append(&mut res);
     }
 
     Ok((result, further_results))
 }
 
+fn execute_batch_search(
+    segment: &LockedSegment,
+    vectors_batch: &Vec<&[VectorElementType]>,
+    prev_params: &BatchSearchParams,
+    use_sampling: bool,
+    total_points: usize,
+    is_stopped: &AtomicBool,
+) -> CollectionResult<(Vec<Vec<ScoredPoint>>, Vec<bool>)> {
+    let locked_segment = segment.get();
+    let read_segment = locked_segment.read();
+    let segment_points = read_segment.available_point_count();
+    let top = if use_sampling {
+        let ef_limit = prev_params
+            .params
+            .and_then(|p| p.hnsw_ef)
+            .or_else(|| get_hnsw_ef_construct(read_segment.config(), prev_params.vector_name));
+        sampling_limit(prev_params.top, ef_limit, segment_points, total_points)
+    } else {
+        prev_params.top
+    };
+    if prev_params
+        .params
+        .map(|p| p.ignore_plain_index)
+        .unwrap_or(false)
+        && read_segment.segment_type() == SegmentType::Plain
```

Comment:
I think this condition might be improved:

- First, the SegmentType::Special (Proxy segments) should be handled differently, depending on the type of currently optimized segment. If it is plain, then we should ignore the segment as well
- Second, It might make sense to allow searching in small segments below indexing threshold. Otherwise, there are no guarantee that small segments will ever be indexed and therefore some data might never become available for search in this mode.

Reply:
makes sense. For 1, this is what I come up with:
    if prev_params
        .params
        .map(|p| p.ignore_plain_index)
        .unwrap_or(false) {
        if match segment {
            LockedSegment::Original(s)  => s.read().segment_type == SegmentType::Plain
            LockedSegment::Proxy(s) => s.read().wrapped_segment.get().read().segment_type() == SegmentType::Plain
        } {
            log::debug!("ignore plain index search:{:?}", read_segment.info());
            let batch_len = vectors_batch.len();
            return Ok((vec![vec![]; batch_len], vec![false; batch_len]));
        }
    }

For 2, any hint as to how to get the indexing_threshold?

Reply:
also under what situation would there be segments < indexing_threshold? Is it the only scenario where the ingestion stops and the last segment is < indexing_threshold? i.e., in non-stop streaming scenario this won't happen?

Reply:
> also under what situation would there be segments < indexing_threshold?

If you insert faster than optimizer can build index

Reply:
>  Is it the only scenario where the ingestion stops and the last segment is < indexing_threshold?

It is also possible with non-stopping streaming, qdrant will create new segments from time to time if all other segments are optimized

Reply:
how to get the indexing_threshold though? I see segment_optimizer has a OptimizerThresholds but I can't find a way to it from segment.